pkg/api/metrics.go

```diff
@@ -97,6 +111,46 @@ func (hs *HTTPServer) QueryMetricsV2(c *models.ReqContext, reqDTO dtos.MetricReq
 	return toMacronResponse(resp)
 }
 
+type queryDSInfo struct {
+	isExpr bool
+	ds     *models.DataSource
+	err    error
+	uid    string
+	id     int64
```

Comment:
:vomiting_face:  -- I don't like this, and it is only required so we can have the info for:
```
return hs.handleGetDataSourceUIDError(q.err, q.uid)
```

An alternate approach would be to have the body of `getDataSourceFromQuery` inside the loop and return errors directly.  Since it is only used once, that could be fine